Use Welford's algorithm for layernorm (DRAM interleaved)

rmillerTT opened the pull request:
### Ticket
#26077 

### Problem description
Use Welford's online algorithm for accumulation mean and variance in layernorm.

### What's changed
Affects only layernorm (small and large variations). This PR is for DRAM interleaved (sharded will be tracked separately). Large variation was reduced from three passes to two passes. A program config parameter was added to use Welford's algorithm for ttnn.layer_norm.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes https://github.com/tenstorrent/tt-metal/actions/runs/17859136486
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/17859144808
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable)
- [ ] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [x] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main) https://github.com/tenstorrent/tt-metal/actions/runs/17814329862
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

vsureshTT committed:
moving machines and saving changes

vsureshTT committed:
broken state

rmillerTT committed:
Adapt layernorm compute kernel for Welford's. Put old kernel in legacy folder for easy removal later

rmillerTT committed:
Rename tile size parameter for clarity

rmillerTT committed:
Initialize Welford

rmillerTT committed:
Add binary op init that was accidentally removed

rmillerTT committed:
Conditionally compile the Welford compute kernel based on ifdef

rmillerTT committed:
Include the Welford header

rmillerTT committed:
Fix typo, remove comment

rmillerTT committed:
[wip] Infra for large-tensor variant of layernorm

rmillerTT committed:
Use sigma in comment again

rmillerTT committed:
Use an env var instead of ifdef to invoke Welford layernorm

rmillerTT committed:
Fix control flow issue and rename env var

rmillerTT committed:
Remove code that was used for testing

rmillerTT committed:
Correct for 1-based indexing

rmillerTT committed:
Make large tensor layernorm/rmsnorm variant a 2-pass algorithm instead of 3

rmillerTT committed:
Comment

rmillerTT committed:
Accumulate x^2 into (\sum x^2)/n for RMS norm

vsureshTT committed:
made copy worl

rmillerTT committed:
Remove bad import

rmillerTT committed:
Factor out test logic and set up pytest

rmillerTT committed:
Use unique CB index for x^2 (bug fix)

rmillerTT committed:
Round out the testing for small layernorm and rms

rmillerTT committed:
Add reconfigs after inits (workaround for a bug)

rmillerTT committed:
Adapt layernorm compute kernel for Welford's. Put old kernel in legacy folder for easy removal later

rmillerTT committed:
Clean out prints

rmillerTT committed:
Use if constexpr instead of preprocessor macros in layernorm

vsureshTT committed:
broken state

rmillerTT committed:
Clean up comment

rmillerTT committed:
Fix a bad merge

rmillerTT committed:
Proper device re-init after reduce op (+ cleanup)

rmillerTT committed:
Small cleanup

rmillerTT committed:
Add tests for RMS norm (with and without fused pre-add)

vsureshTT committed:
moving machines and saving changes

rmillerTT committed:
Add some tests for various layernorm code paths using the dummy mean and variance

vsureshTT committed:
moving machines and saving changes

vsureshTT committed:
broken state

rmillerTT committed:
Adapt layernorm compute kernel for Welford's. Put old kernel in legacy folder for easy removal later

vsureshTT committed:
made copy worl

djordje-tt committed:
Revert "Topk: parallel over more cores in sub range" (#26770)

Reverts tenstorrent/tt-metal#24517
Breaks some tests.
@djordje-tt

djordje-tt committed:
Revert "Topk: parallel over more cores in sub range" (#26770)

Reverts tenstorrent/tt-metal#24517
Breaks some tests.
@djordje-tt

rmillerTT committed:
Fix merge issue

rmillerTT committed:
Transpose through CB interface

rmillerTT committed:
Fix bug

rmillerTT committed:
Get large tensor layernorm working (will clean up)

djordje-tt committed:
TopK larger core range (#26840)

Due to reverted original PR.
Revert:
https://github.com/tenstorrent/tt-metal/commit/685cb14221e5f200694bd9f957ea8228b2dcf298
Original PR: https://github.com/tenstorrent/tt-metal/pull/24517
On the top of original changes only added topk_utils.cpp file to split
definition and implementation to fix linking bug.

- [ ] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/runs/16960036070)
- [ ] [6U CI
pipelines](https://github.com/tenstorrent/tt-metal/actions/runs/16959960716)

---------

Co-authored-by: Johanna Rock <jrock@tenstorrent.com>

djordje-tt committed:
TopK larger core range (#26840)

Due to reverted original PR.
Revert:
https://github.com/tenstorrent/tt-metal/commit/685cb14221e5f200694bd9f957ea8228b2dcf298
Original PR: https://github.com/tenstorrent/tt-metal/pull/24517
On the top of original changes only added topk_utils.cpp file to split
definition and implementation to fix linking bug.

- [ ] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/runs/16960036070)
- [ ] [6U CI
pipelines](https://github.com/tenstorrent/tt-metal/actions/runs/16959960716)

---------

Co-authored-by: Johanna Rock <jrock@tenstorrent.com>

rmillerTT committed:
Get large tensor LN working (no fused add or gamma/beta yet)

rmillerTT committed:
More large tensor work

rmillerTT committed:
Factor out the pack/transpose of ex and ex2 combined into one function

rmillerTT committed:
Get fused residual working for large layernorm (gamma/beta also working)

rmillerTT committed:
Get large RMS working

rmillerTT committed:
Small change to reconfig to make cb's consistent

rmillerTT committed:
Restore kernels previously put in _legacy folders back to regular kernel folders, rename welford's kernels, and update control logic to reference these new files

rmillerTT committed:
Hook up Welford layernorm flag to program config

rmillerTT committed:
Remove the function used for testing

rmillerTT committed:
Remove the test_welford.py file and expand the test_layer_norm test suite

rmillerTT committed:
Fix assert for RMS norm path

rmillerTT committed:
Fix test params

rmillerTT committed:
Reserve before pack

rmillerTT committed:
Use add tiles instead of add tiles dest, and use full transpose init

rmillerTT committed:
Update skip condition on test

rmillerTT committed:
Remove comment

bbradelTT commented on ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/dataflow/reader_unary_interleaved_ln_large_tensor_welford.cpp:
```diff
@@ -0,0 +1,117 @@
+template <bool DRAM, uint32_t tile_hw = 1024>
+void read_block_to_cb(
+    const uint32_t cb_id,
+    const InterleavedAddrGenFast<DRAM, tile_hw>& addr,
```
You need to use TensorAccessor. See https://github.com/tenstorrent/tt-metal/pull/26916

AI is pretty good at updating the code, although manually is probably okay too.

If that PR is merged first, then update here. Otherwise do a follow up PR.

bbradelTT commented on ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/compute/layernorm_welford.cpp:
```diff
@@ -0,0 +1,329 @@
+
+    }  // NCHt loop
+    // cb_pop_front(cb_scaler, 1); // optional for correctness
+    // cb_pop_front(cb_eps, 1); // optional for correctness
```
Should be removed.

bbradelTT commented on models/demos/llama3_70b_galaxy/demo/text_demo_targets.json:
```diff
@@ -3,7 +3,7 @@
             "prefill_pcc": 0.9125991571810271,
             "decode_pcc": 0.9498023316321954,
-            "throughput": 61.00,
+            "throughput": 59.87,
```
This should not have changed.

Why is it different?

bbradelTT commented on ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/compute/layernorm_welford.cpp:
```diff
@@ -0,0 +1,329 @@
+    constexpr uint32_t onetile = 1;
+    constexpr bool layernorm = !rms_norm;
+    // reserve one tile for zeros on cb_in2
+    // TODO(AP): check that if DST is indeed zeroed by release_dst (and initially), we can use it as zeroes
```
Please put in an issue number if this is not done as part of this PR.

rmillerTT committed:
Remove unnecessary comments

rmillerTT committed:
Revert "Revert "Topk: parallel over more cores in sub range" (#26770)"

This reverts commit 82738f5a345a45ca382110cacdb8e2150b177c89.

rmillerTT commented:
Having merge issues atm, this PR is a bit butchered. Fixing now

rmillerTT committed:
Revert "Factor out test logic and set up pytest"

This reverts commit 5c6e2554745f2ea4ff3b11f57e4952eaba51aef7.

rmillerTT committed:
This was broken in the merge, fixing

rmillerTT committed:
Use TensorAccessor for large Welford's

bbradelTT commented on tests/ttnn/unit_tests/operations/fused/test_layer_norm.py:
```diff
@@ -14,14 +14,16 @@
 @pytest.mark.parametrize("h", [32])
 @pytest.mark.parametrize("w", [64])
-def test_layer_norm(device, h, w):
+@pytest.mark.parametrize("use_welford", [True, False])
```
Welford's would need to be skipped on BH until the kernel is ported.

rmillerTT committed:
Skip the Welford tests on BH

vsureshTT committed:
some LLK cleanup

vsureshTT committed:
updated llk

rdjogoTT committed:
Change binary/ternary SFPU API and kernels to use tile idx directly (#27513)

https://github.com/tenstorrent/tt-metal/issues/19442
https://github.com/tenstorrent/tt-llk/issues/580

We used to take the min(tile_idx0, tile_idx1, ...) and program the Dest
math offset to that in order to perform SFPU kernels by passing down
just the "offset" between operands. This broke when the second operand
was located in an earlier tile_idx than the first for non-commutative
OPs. Also, the tile at offset 0 was always used as output with no
flexibility.

For both BH and WH:
- change the API and kernels to accept an idx for each tile used (input
tiles and output tile) and update the kernels accordingly
- update all binary sfpu calls in metal to have an out_tile (set to same
as input tile 0 arg since that was the previous default)

- [x] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/runs/17303734096)
CI passes
- [x] [Blackhole Post
commit](https://github.com/tenstorrent/tt-metal/actions/runs/17303737605)
CI with demo tests passes (if applicable)
- [x] [Model
regression](https://github.com/tenstorrent/tt-metal/actions/runs/17275855237)
CI passes (if applicable)
- [x] [Device performance
regression](https://github.com/tenstorrent/tt-metal/actions/runs/17275857042)
CI passes (if applicable)

rdjogoTT committed:
Change binary/ternary SFPU API and kernels to use tile idx directly (#27513)

https://github.com/tenstorrent/tt-metal/issues/19442
https://github.com/tenstorrent/tt-llk/issues/580

We used to take the min(tile_idx0, tile_idx1, ...) and program the Dest
math offset to that in order to perform SFPU kernels by passing down
just the "offset" between operands. This broke when the second operand
was located in an earlier tile_idx than the first for non-commutative
OPs. Also, the tile at offset 0 was always used as output with no
flexibility.

For both BH and WH:
- change the API and kernels to accept an idx for each tile used (input
tiles and output tile) and update the kernels accordingly
- update all binary sfpu calls in metal to have an out_tile (set to same
as input tile 0 arg since that was the previous default)

- [x] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/runs/17303734096)
CI passes
- [x] [Blackhole Post
commit](https://github.com/tenstorrent/tt-metal/actions/runs/17303737605)
CI with demo tests passes (if applicable)
- [x] [Model
regression](https://github.com/tenstorrent/tt-metal/actions/runs/17275855237)
CI passes (if applicable)
- [x] [Device performance
regression](https://github.com/tenstorrent/tt-metal/actions/runs/17275857042)
CI passes (if applicable)

rmillerTT committed:
[wip] Merge with welford changes, rewrite large welford LN kernel to be more in line with programming model

rmillerTT committed:
Reworked large Welford LN (working). Will add RMS back in next

rmillerTT committed:
Put the RMS path back in (note: This is untested, as Welford's for RMS is currently not implemented, and the factory uses a different kernel for large-tensor RMS)

rmillerTT committed:
Get rid of prints

rmillerTT committed:
Get rid of prints

rmillerTT committed:
Fix the small Welford LN tests

rmillerTT committed:
Remove untested RMS LN Welford's code (but leave the plumbing for if and when it does become supported). Currently fatal'd during program construction

rmillerTT committed:
Restore old test parameters

rmillerTT committed:
Merge in the welford API changes

rmillerTT committed:
Fix merge issues

rmillerTT committed:
Pass false to M2 to var (this shouldn't need to happen, but welford LN tests fail unless I do this). It's a bug that needs investigation

rmillerTT committed:
Revert "Change binary/ternary SFPU API and kernels to use tile idx directly (#27513)"

This reverts commit 69da853d64c45e2b868c37b3a5337d751411f5d9.

rmillerTT committed:
Fix rebase issue

rmillerTT committed:
Expose the LayerNormDefaultProgramConfig default constructor through pybind

rmillerTT committed:
Remove unused variable definition

rmillerTT committed:
Remove welford header from non-welford kernel

rmillerTT committed:
Fix merge issue

rmillerTT committed:
Add a fatal for restricting welford LN to WH

rmillerTT committed:
Remove RMS logic from welford paths (RMS always uses legacy layernorm)

rmillerTT committed:
Merge with new legacy reduction/rsqrt options

rmillerTT committed:
Use rsqrt in the welford LN kernels

rmillerTT committed:
Merge with the LLK changes

rmillerTT committed:
Remove sfpu2 file that should not have been there

rmillerTT commented:
Model perf, device perf, and demo tests all have 1-2 failures that are present in either main (the first two) or another unrelated branch (demo tests).

Pipelines have been unstable over the last few days and many attempts were made at getting all pipelines green for this branch, but it's time to merge this
